core/include/traccc/bfield/magnetic_field.hpp

```diff
@@ -64,12 +65,55 @@ class magnetic_field {
     template <covfie::concepts::field_backend bfield_backend_t>
     const covfie::field<bfield_backend_t>& as_field() const;
 
+    /// @brief Return type information about the contained magnetic field.
+    const std::type_info& type() const { return m_field.type(); }
+
     private:
     /// The actualy covfie b-field object
     std::any m_field;
 
 };  // class magnetic_field
 
+/// @brief Helper function for `bfield_visitor`
+template <typename callable_t, typename bfield_t, typename... bfield_ts>
+auto magnetic_field_visitor_helper(const magnetic_field& bfield,
+                                   callable_t&& callable,
+                                   std::tuple<bfield_t, bfield_ts...>*)
+    requires(covfie::concepts::field_backend<bfield_t> &&
+             (covfie::concepts::field_backend<bfield_ts> && ...))
+{
+    if (bfield.is<bfield_t>()) {
+        return callable(bfield.as_view<bfield_t>());
+    } else {
+        if constexpr (sizeof...(bfield_ts) > 0) {
+            return magnetic_field_visitor_helper(
+                bfield, std::forward<callable_t>(callable),
+                static_cast<std::tuple<bfield_ts...>*>(nullptr));
+        } else {
+            std::stringstream exception_message;
+
+            exception_message
+                << "Invalid B-field type (" << bfield.type().name()
+                << ") received, but this type is not supported" << std::endl;
+
+            throw std::invalid_argument(exception_message.str());
+        }
+    }
+}
+
+/// @brief Visitor for polymorphic magnetic field types
+///
+/// This function takes a list of supported magnetic field types and checks
+/// if the provided field is one of them. If it is, it will call the provided
+/// callable on a view of it and otherwise it will throw an exception.
+template <typename bfield_list_t, typename callable_t>
+auto magnetic_field_visitor(const magnetic_field& bfield,
+                            callable_t&& callable) {
+    return magnetic_field_visitor_helper(bfield,
+                                         std::forward<callable_t>(callable),
+                                         static_cast<bfield_list_t*>(nullptr));
```

Comment:
Would [std::declval](https://en.cppreference.com/w/cpp/utility/declval.html) not be more elegant here?

Reply:
Never mind, apparently not. 🤔